src/main/java/dev/blaauwendraad/masker/json/MaskingState.java

```diff
@@ -66,9 +92,26 @@ public byte[] getMessage() {
      * @see ReplacementOperation
      */
     public void replaceTargetValueWith(int startIndex, int length, byte[] mask, int maskRepeat) {
-        ReplacementOperation replacementOperation = new ReplacementOperation(startIndex, length, mask, maskRepeat);
-        replacementOperations.add(replacementOperation);
-        replacementOperationsTotalDifference += replacementOperation.difference();
+        if (outputStream == null) {
+            ReplacementOperation replacementOperation = new ReplacementOperation(startIndex, length, mask, maskRepeat);
+            replacementOperations.add(replacementOperation);
+            replacementOperationsTotalDifference += replacementOperation.difference();
+        } else {
+            // write the replacement into the output stream
+            try {
+                // write everything up to the beginning of the current replacement
+                outputStream.write(message, lastReplacementEndIndex, startIndex - lastReplacementEndIndex);
+
+                // write the replacement
+                for (int i = 0; i < maskRepeat; i++) {
+                    outputStream.write(mask);
+                }
+                outputStream.flush();
```

Comment:
Do we actually need `flush` here?

Reply:
Not really. Now as I think of it, a single flush while reloading the buffer makes more sense.

Reply:
Done